rfcs/20200712-tfrt-kernel-fallback.md

```diff
@@ -0,0 +1,646 @@
+[current one](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/kernels.cc) -
+implemented in terms of TensorFlow runtime. Second implementation will use TFRT
+Kernel Fallback instead. We can select between the two kernel C API
+implementations by adding a build config setting:
+
+```
+# Whether to use TFRT-based implementation of the kernel C API.
+config_setting(
+    name = "tfrt_kernel_c_api",
+    define_values = {
+        "tfrt_kernel_c_api": "True",
+    },
+)
+```
+
+Most of the kernel C API implementation will be the same between the two with a
+few notable exceptions:
+
+*   TFRT Kernel Fallback implementation will cast `TF_OpKernelContext` and
+    `TF_OpKernelConstruction` to `TFRTOpKernelContext` and
+    `TFRTOpKernelConstruction` respectively.
+*   TFRT Kernel Fallback implementation will use Kernel Fallback registration
+    mechanism.
+
+### TFRT forwarding kernel registration using C API
+
+We plan to implement C API for TFRT kernel registration that calls TFRT Kernel
+Fallback registration mechanism. Note that this is analogous to TF Lite
+providing
+[their own C API registration mechanism](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/lite/c/common.h;l=739?q=tfliteregistration&ss=tensorflow%2Ftensorflow).
+
+```cpp
+TF_KernelBuilder* TF_NewKernelBuilder(
+    const char* op_name, const char* device_name,
+    void* (*create_func)(TF_OpKernelConstruction*),
+    void (*compute_func)(void*, TF_OpKernelContext*),
+    void (*delete_func)(void*)) {
+  TF_KernelBuilder* result = new TF_KernelBuilder;
+  result->create_function = create_func;
+  result->compute_function = compute_func;
+  result->delete_function = delete_func;
+  return result;
+}
+
+void TF_RegisterKernelBuilder(const char* name,
+                              TF_KernelBuilder* builder,
+                              TF_Status* status) {
+  auto* create_fn = builder->create_function;
+  auto* compute_fn = builder->compute_function;
+  auto* delete_fn = builder->delete_function;
+  auto create_kernel = [create_fn, compute_fn, delete_fn](
+      TFRTOpKernelConstruction* construction) {
+    return std::make_unique<tensorflow::TFRTCOpKernel>(
+        construction, create_fn, compute_fn, delete_fn);
+  };
+  ::tensorflow::TFRTKernelReg kernelinfo(create_kernel);
+  kernelinfo.type_constraints = builder->attr_to_type;
+  ::tensorflow::tfrt_kernel_factories->RegisterFactory(
+      name, kernelinfo);
+  tensorflow::TFRTOpRegisterer(tensorflow::TFRTOpMetaBuilder(name));
+  TF_DeleteKernelBuilder(builder);
+  TF_SetStatus(status, TF_OK, "");
+}
+```
+
+## TFRT integration
+
+Current preferred direction would generate a
+[BEF](https://github.com/tensorflow/runtime/blob/master/documents/binary_executable_format.md) file in advance and then run that
+file on a mobile device. Generated BEF file would have to call either native, TF
+Lite, runtime fallback or kernel fallback kernels and provide any glue logic in
```

Comment:
I think it would be good to clarify here that when the reference to "TF Lite" is the current TF Lite runtime via TFLite Select, as opposed to TFRT running on mobile which is the preferred longer term path.

Reply:
I want to understand the difference in supporting one or the other better. Ideally, we want to support both.

Reply:
TFLite and TFRT are separate pieces of infra that, as you mention, will benefit from your work here.
TFLite is the current TF mobile solution, whereas we're working on making TFRT a viable solution for mobile as well.
  
I just mean that to the reader not familiar with these details may get confused by the references to TFLite, and TFRT, and mobile.

Maybe clarifying this further in the Goals section would be good.

Reply:
@miaout17 
Also, we're not planning on BEF calling TFLite kernels.

Reply:
I added a clarification in Goals that this document is focusing on running kernels from TFRT while running on mobile. Exact details how TFRT itself is executed on mobile is outside of the scope here.
I also removed reference to TF Lite kernels in this paragraph.